File: InterviewProblemsSolving/Python/coding_problem_1.py
"""
Beginners basic math probs - Find even or odd
Problem link : https://www.geeksforgeeks.org/problems/odd-or-even3618/1
To find whether the number is even or odd.
"""

# Method 1: Using modulus.
print("Method 1: Using modulus.");

def is_even_modulus(number):
    return number % 2 == 0;
# If true then it is even.
# If false then it is false.
print(is_even_modulus(6));
print(is_even_modulus(9));

# Method 2: Using binary.
print("Method 2: Using binary.");

# The main concept here is binary even numbers end with 0 so even & 1 is 0, odd numbers end with 1 so odd & 1 is 1.
def is_even_binary(number):
    return number & 1 == 0;
print(is_even_binary(6));
print(is_even_binary(9));

# Method 3: Using Bitwise Shift.
print('Method 3: Using Bitwise Shift.');
# Bitwise shift rule.
# For even numbers when we shift once right and once left the initial and final binary number remains the same.
# For odd numbers when we shift once right and once left the initial and final  binary number is different.
def is_even_bitwise(number):
    return number == (number >> 1) << 1 ;
print(is_even_bitwise(6));
print(is_even_bitwise(9));

File: InterviewProblemsSolving/JavaScript/coding_problem_1.js
// Find odd or even.

console.log('Method 1: Using Modulus');
// Method 1: Using Modulus
function isEvenModulus(number){
    return (number % 2) == 0;
}

console.log(isEvenModulus(6));
console.log(isEvenModulus(9));

// Use arrow function
const isEvenModulusArrow = (number) => number % 2 == 0;

console.log(isEvenModulusArrow(6));
console.log(isEvenModulusArrow(9)); 

console.log('Method 2: Using Binary');
function isEvenBinary(number){
    return (number & 1) == 0;
}
console.log(isEvenBinary(6));
console.log(isEvenBinary(9)); 
// Arrow function.
const isEvenBinaryArrow = (number) => (number & 1) == 0;
console.log(isEvenBinaryArrow(6));
console.log(isEvenBinaryArrow(9)); 

console.log('Method 3: Using Bitwise shift.');
function isEvenBitwiseShift(number){
    return number == (number >> 1) << 1;
}
console.log(isEvenBitwiseShift(6));
console.log(isEvenBitwiseShift(9)); 
// Arrow function.
const isEvenBitwiseShiftArrow = (number) => number == (number >> 1) << 1;
console.log(isEvenBitwiseShiftArrow(6));
console.log(isEvenBitwiseShiftArrow(9)); 